README
			 GOCR (JOCR at SF.net)

GOCR is an optical character recognition program, released under the
GNU General Public License. It reads images in many formats  and outputs
a text file. Possible image formats are pnm, pbm, pgm, ppm, some pcx and
tga image files. Other formats like pnm.gz, pnm.bz2, png, jpg, tiff, gif,
bmp will be automatically converted using the netpbm-progs, gzip and bzip2
via unix pipe.
A simple graphical frontend written in tcl/tk and some
sample files are included.
Gocr is also able to recognize and translate barcodes.
You do not have to train the program or store large font bases.
Simply call gocr from the command line and get your results.

To see installation instructions, see the INSTALL file.

How to start? (QUICK START)
---------------------------
Some examples of how you can use gocr:

  gocr -h       			# help
  gocr file.pbm				# minimum options
  gocr -v 1 file.pbm >out.txt 2>out.log # generate text- and log file
  djpeg -pnm -gray text.jpg | gocr -	# using JPEG-files
  gzip -cd text.pbm.gz | gocr -		# using gzipped PBM-files
  giftopnm text.gif | gocr -		# using GIF-files
  gocr -v 1 -v 32 -m 4 file.pbm   	# zoning and out30.bmp output
  xloadimage -geometry 400x400 out30.png # see details using image viewer
  gocr -f XML -i file.pgm -o file.xml    # output simple XML-format
  wish gocr.tcl			  # X11-tcl/tk-frontend (development version)
  # see manual pages for more details


How to get image files?
-----------------------
Scan text pages and save it as PGM/PBM/PNM file. Use a program such as
The GIMP or Sane. You can also use netpbm-progs to convert several image 
formats into PGM/PBM/PNM. The tool djpeg can be used to convert jpeg into pgm.
If you have a POSIX compatible system like linux and PNM-tools, gzip and bzip2
are installed, you are lucky and gocr will do conversion 
from [.pnm.gz, .pnm.bz2, .jpg, .jpeg, .bmp, .tiff, .png, .ps, .eps]
to [.pgm] for you. This list can easily be extended editing src/pnm.c.

Gocr also comes with some examples, try: make examples.

Memory limitations
------------------
WARNING!!!

If you use a 300dpi scan of A4 letter, the image is about 2500x3500 pixels and
gocr requires 8.75MB for storing the picture into the memory. Not only that,
but gocr may create a 2nd copy, using a total of 17MB. This is independent
of using b/w or gray-scale images. Be sure that you have enough RAM installed
in your machine! Alternatively you can cut the picture into small pieces.
You can use the pnmcut, from the netpbm package to cut the file. Example:

pnmcut -left 0 -right 2500 -top 0 -height 1000 bigfile.pnm > smallfile.pnm

And then use gocr in the cropped image as usual. Take care: if you chopp the
characters, gocr won't be able to understand that line.

Future versions will take care of this issue automatically.

Limitations
-----------
gocr is still in its early stages. Your images should fit in these requirements
if you want a good output:

- good scans (all chars well seperated, one column, no tables etc, 12pt 300dpi)
  should work well
- fonts 20-60 pixels ( 5pt * 1in/72pt * 300 dpi = 20 dots )
- output of image file for controlling detection

And note that speed is very slow (this will be changed when recognition works 
well)
  12pt 300dpi 1700x950 16lines 700chars 22x28 P90=40s..90s v0.2.3 (gcc -O0)

You can try to optimize the results:
- make good scans/treat image
- try to change the critical gray level (option -l <n>)
- control the result on out10.png, out30.png (option -v 32)
  example: ./gocr -v 32 -m 4 -m 256 -m 56 ~/aac.jpg # only check layout
- enlarge option -d <n> for high resolution images which are noisy
- try different combinations for option -m <n>
- for thousends of documents with same font
  you can use/create a database (-m 2/-m 130)
- use options -d 0 -m 8 on screen shots (font8x12)
- use filter option -C to through out wrong recognized chars (ex: gothic)

What does >> NOT << work at the moment:
- complex layouts (try option -m 4)
- bad scans, noisy/snowy images, FAX-quality images
- serif fonts, italic fonts, slanted fonts
- handwritten texts (this is valid for the next ten years I guess)
   the exisctence of autotrace can shorten this 
- rotated images (but slightly rotated images should be no problem)
- small fonts (fax like) or mix of different font size
- colored images (use gray or black/white)
- Chinese, Arabian, Egyptian, Cyrillic or Klingon fonts

How it works or how it should work?
- put the entire file into RAM (300dpi grayscale recommended)
- remove dust and snow
- detect small angle (lines which are not horizontal)
- detect text boxes (option -m 4)
- detect text-lines
- detect characters
- first step recognition (every character has its own empirical procedure)
  - no neural network or similar general algorithms
- analyze not detected chars by comparison with detected ones
- try to divide overlapping chars
- testwise: compare all letters (like compression of pictures)
- for more details look to the gocr.html documentation

Why the result of the new version are worse compared to the old version?
- the algorithms of gocr are sometimes evolutionary, a fine tuned old
  algorithm will be replased by a completly new algorithm which is more
  general but a bit worse for your problem.
  Please send your sample and give the new algo the chance to become
  better as the old one.

Security
--------
Because gocr only reads and writes files it is quite sure, except the
popen-function which allows you to call gocr with non-pnm-image formats
directly. The popen function can be misused to start other probably 
dangerous programs.
If you care about conversion to pnm format, you can safely disable
popen-function by removing "#define HAVE_POPEN 1" from config.h before
compiling the gocr package.


How can you help gocr?
----------------------
- Send comments, ideas and patches (diff -ru gocr_original/ gocr_changed/).
- If you found a bug, i.e. clear characters not recognized,
  crop the area around the problem of about 3 text lines and
  send as format png or jpeg for fotos, small images are easier
  for debugging and will go to my check-database.
- I always need small example files (.pbm.gz, png or jpeg)
  of maximum 100kB for testing
  the behavior of the ocr engine under different conditions, 
  because scanning does take a lot of time which I do not have.
  Please only use free image formats: .pbm.gz (b/w), png (screenshots,
  converted pdfs) or jpeg (optical scans, photos).
  That will help, to get the world's best OCR open source program. :) Thanks!
- Please dont send captchas. GOCR is mainly intended to make books 
  and knowledge easier accessible.
- If you have a good idea, how to manage some OCR-tasks, tell me!
- If you have a lot of money, spend a bit (paypal). Ok, paypal has changed,
  so please forget about it. Also the importand problem is now the missing
  spare time for coding.


After all, is it gocr or jocr?
------------------------------
The original name of this project is gocr, from GPLed Optical Character
Recognition. Another project is using the same name, however; so the
name was changed to jocr. If you have a good idea for a name, please
send it.


Latest news
------------
  http://www-e.uni-magdeburg.de/jschulen/ocr/

Authors: (see AUTHORS)


HISTORY
History: (Changes,ChangeLog)

 0.53pre 
   2019-04 fix Makefile.in some prefix/DESTDIR mixture
   2019-04 fix thresholding valgrind.memcheck + exchange cols rows arguments
   2019-04 add PAM support, fix bad PNM format handling, fix multi-PNM

 0.52 Sep18
   2018-10 fix endianess of 16bit-pnm (NetPBM: most significant byte first)
           it works partly with (old) wrong order, but noisy contour
   2018-09 improve tests: random | base64 as FreeMono-Regular 80pt
   2018-09 fix bad 7 as T detection, fix corner vectors for thinn fonts
           fix debug-option-dependence -v32 of iIl|-vert.-line-detection
               other chars may have that problem also
   2018-09 skip UTF8 code above 16bit if __WCHAR_MAX__ is 16bit (VS)
   2018-09 simplify xml-format, add xml-sample to the README
           achars (alternative chars) include the main char now
   2018-09 fix reading P5-PGM/PNM format (pnmtoplainpnm)
   2018-09 some clean up compiler warnings, set default --with-debug
   2018-09 error + exit on bad option, fix missing -h for --help

 0.51 Jun13-Aug17
   2017-08 fix some 8x9 (unsharp) screen fonts (0O,il1I,e),
           from old samples and patches received 2005
   2017-04 fix NULL-pointer access by Norbert M.
   fix range check in nearest_frame_vector() (does not affect users)
   add appended argument to option ("-v33" or "-v 33")
   fix J vs. 3 (13x20)
   fix compiler warning by typo, thx to Senh Liu, Jun2013
   (still lot on my todo list)
   
 0.50 Sep10-May12,Mar13
   just release it to avoid questiions to old problems, give a life sign ;)
   fix 4 parfait problems against 0.48 (thanks to Rich Burridge) 
   adding qrcode detection and decoding (no error correction, no skewing)
   spacing slightly improved
   context correction of hex codes (p.e. hex fingerprints)
   some threshold value adaptions (not finished)
   try to fix double output of XML code <...> and removed additional \n
   improved quotation detection ,, ''
   improved monospaced spacing (video text)

 0.49 Aug09-Sep10
   fix dot handling for ':' and ';' (vector code)
   fix '@' for 7x9 and 5x8 fonts
   fix double counting of subboxes (affects "0" (zero) with dot in it)
   character "l" of width 1 improved
   bug fix gluing chars ij of width=1
   bug fix thresholding (small gray images)
   return error code -1 on ERROR pnm.c unexpected EOF
   fix conflicts with unicode_defs.h TRUE definition on gcc/alphaev7-osf/3.4
   further fixes for lib by D. Katsubo
   fix #3039007 "struct list" in list.h conflicts with STL (ocr_object_list)
   fix #3039006 INFINITY macro in unicode.h conflicts with math.h
   bugfix barcode 128, switch from mode mC to mA (":1")
   bugfix: MultiPNM + database  - ID: 2957140
   improved barcode recognition - ID: 2859644 (bars wider than spaces)
   quality test-script bin/gocr_chk.sh added
   initial datamatrix support (ASCII + ASCII numeric only, no ErrCorrection)

 0.48 Jul09
       fix buffer overflow introduced in 0.46 for filenames
       add codabar barcode
       fix bug, removing melted serifs
       add patch by Chris Lee, i25 barcode recognition + modifications
       fix some false positive numbers "34" (video, gas meter)
       fix problems with 2zZ4 for 10x10 screen font
       better debug output for :;,.
       remove examples, doc and libs part from configure (see below)
       remove doc and examples from the (make install) part to reduce
         dependencies (gs and transfig is not needed for rpm/ebuild)
         gocr only may depend from netpbm, but can live without too
         this will help to install gocr on "exotic" (nonlinux) platforms
       fix gentoo app-text/gocr Bug 243250 src/Makefile: $(CC) $(LDFLAGS) ...

 0.47  fix database recognition for certainty 100 (-a 100)
       insert spaces with certainty 100 (old: 99) to let -a 100 work
       new option -u string for unrecognized chars
       fix: No contrast in image causes division by zero
       reduced false positive recognition of scanned "a496" (Gutenberg Project)
       "d as a" patch ID: 1556112
       add "Windows Pipe Fix", but I hate extra code for bad environments
       improve 7x10, sample 0811qemu1.png (ToDo: not finished)
       change black:white from &gt;4:1 to &gt;3.5:1 as criteria of inversion
       reintroduce static library libPgm2asc.a (make libs) for OSRA project
       add dynamic library (make libs), unused but may help other projects
      
 0.46  improved context correction (especially helvetica "Il")
       improved recognition of tiny chars "$1", fat "s", "rw" ","
       fix blank spaces problem in filenames
         (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=316511)
         !!! please check on other platforms and report to me !!!
         there are still problems with special chars (double quodes, backslash)
         better use this way: djpeg -gray -pnm strangefilename.jpg | gocr -
       fix possible problem with database and UTF8 input 
       fix hidden bug in pitch/spacing initialization
       reactivate code for output of glued chars and strings 
       fix wrong close() call
       remove creation of pgm2asc.a for simplicity (see SF-patch 1827477)
       
 0.45  minor corrections for c and k
       minus sign is filtered by option -C "--" now, ("\-" was parsed badly)
       clean up old unused code for simplicity (api, frontend)
       fix problem with low height barcodes and barcode removing
       fix problem with readpgm (for multiple images) and database
       PACKAGE_VERSION defined by configure.in AC_INIT + gocr.spec

 0.44  add volume to boxes (negative means white areas inside black areas)
       Fix overflow in despeckling routine (verbose mode, dust removing)
       reactivate composed chars, fix merge_boxes
       fix problems with uncertain line detection and not recognized "7"
       option -a has an effect now for the output
       adaptions to MICR E13-B font (see GnuMICR), ToDo: 4 extra-chars
       fix num_boxes in merge_boxes (affects line detection)

src/gocr.c
/*
This is a Optical-Character-Recognition program
Copyright (C) GPLv2 2000-2019  Joerg Schulenburg

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

 see README for EMAIL-address

  sometimes I have written comments in german language, sorry for that

 This file was retrieved from pgm2asc.cc of Joerg, in order to have
 a library of the ocr-engine from Klaas Freitag
 
 */
#include "config.h"
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#ifdef HAVE_GETTIMEOFDAY 
#include <sys/time.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "pnm.h"
#include "pgm2asc.h"
#include "pcx.h"
#include "ocr0.h"		/* only_numbers */
#include "progress.h"
#include "version.h" /* include/version.h version_string + release_string */

/* v+=2 exit, v+=1 add_how_to_get_help */
static void out_version(int v) {
  fprintf(stderr, " Optical Character Recognition --- gocr "
          version_string " " release_string "\n"
          " Copyright (C) 2001-2019 Joerg Schulenburg  GPG=1024D/53BDFBE3\n"
          " released under the GNU General Public License\n");
  /* as recommended, (c) and license should be part of the binary */
  /* no email because of SPAM, see README for contacting the author */
  if (v)
    fprintf(stderr, " use option -h for help\n");
  if (v & 2)
    exit(1);
  return;
}

static void help(void) {
  out_version(0);
  /* output is shortened to essentials, see manual page for details */
  fprintf(stderr,
	  " using: gocr [options] pnm_file_name  # use - for stdin\n"
	  " options (see gocr manual pages for more details):\n"
	  " -h, --help, -V --version\n"
	  " -i name   - input image file (pnm,pgm,pbm,ppm,pcx,...)\n"
	  " -o name   - output file  (redirection of stdout)\n"
	  " -e name   - logging file (redirection of stderr)\n"
	  " -x name   - progress output to fifo (see manual)\n"
	  " -p name   - database path including final slash (default is ./db/)\n");
  fprintf(stderr, /* string length less than 509 bytes for ISO C89 */
	  " -f fmt    - output format (ISO8859_1 TeX HTML XML UTF8 ASCII)\n"
	  " -l num    - threshold grey level 0<160<=255 (0 = autodetect)\n"
	  " -d num    - dust_size (remove small clusters, -1 = autodetect)\n"
	  " -s num    - spacewidth/dots (0 = autodetect)\n"
	  " -v num    - verbose (see manual page)\n"
	  " -c string - list of chars (debugging, see manual)\n"
	  " -C string - char filter (ex. hexdigits: ""0-9A-Fx"", only ASCII)\n"
	  " -m num    - operation modes (bitpattern, see manual)\n");
  fprintf(stderr, /* string length less than 509 bytes for ISO C89 */
	  " -a num    - value of certainty (in percent, 0..100, default=95)\n"
	  " -u string - output this string for every unrecognized character\n");
  fprintf(stderr, /* string length less than 509 bytes for ISO C89 */
	  " examples:\n"
	  "\tgocr -m 4 text1.pbm                   # do layout analyzis\n"
	  "\tgocr -m 130 -p ./database/ text1.pbm  # extend database\n"
	  "\tdjpeg -pnm -gray text.jpg | gocr -    # use jpeg-file via pipe\n"
	  "\n");
  fprintf(stderr, " website: http://www-e.uni-magdeburg.de/jschulen/ocr/\n");
/* fprintf(stderr, " website: http://jocr.sourceforge.net/ (2000-2013)\n"); */
  exit(0);
}

#ifdef HAVE_GETTIMEOFDAY
/* from the glibc documentation */
static int timeval_subtract (struct timeval *result, struct timeval *x, 
    struct timeval *y) {

  /* Perform the carry for the later subtraction by updating Y. */
  if (x->tv_usec < y->tv_usec) {
    int nsec = (y->tv_usec - x->tv_usec) / 1000000 + 1;
    y->tv_usec -= 1000000 * nsec;
    y->tv_sec += nsec;
  }
  if (x->tv_usec - y->tv_usec > 1000000) {
    int nsec = (x->tv_usec - y->tv_usec) / 1000000;
    y->tv_usec += 1000000 * nsec;
    y->tv_sec -= nsec;
  }

  /* Compute the time remaining to wait.
     `tv_usec' is certainly positive. */
  result->tv_sec = x->tv_sec - y->tv_sec;
  result->tv_usec = x->tv_usec - y->tv_usec;

  /* Return 1 if result is negative. */
  return x->tv_sec < y->tv_sec;
}
#endif

static void process_arguments(job_t *job, int argn, char *argv[])
{
  int i;
  char *s1;

  assert(job);

  if (argn <= 1) {
    out_version(1);
    exit(0);
  }
#ifdef HAVE_PGM_H
  pnm_init(&argn, &argv);
#endif
  
  /* process arguments */
  for (i = 1; i < argn; i++) {
    int iskip= 0; /* 2017-03-26 allow -v33 (iskip=0) or -v 33 (iskip=1) */
    if (strcmp(argv[i], "--help") == 0
     || strcmp(argv[i], "-h")     == 0) help(); /* and quits  2018-09 -h */
    if (strcmp(argv[i], "--version") == 0
     || strcmp(argv[i], "-V")        == 0) { /* and quits  2018-09 */
       printf(version_string "-" release_string "\n"); exit(0);}
    if (argv[i][0] == '-' && argv[i][1] != 0) {
      if (i + 1 < argn && argv[i][2]=='\0') { s1 = argv[i + 1]; iskip=1; }
      else if (argv[i][2]!='\0')              s1 = argv[i]+2;
           else {
             fprintf(stderr,"ERROR: missing argument, try --help\n");
             exit(1);
           }
      switch (argv[i][1]) {
      case 'i': /* input image file */
	job->src.fname = s1; break;
      case 'e': /* logging file */
	if (s1[0] == '-' && s1[1] == '\0') {
#ifdef HAVE_UNISTD_H
          dup2(STDOUT_FILENO, STDERR_FILENO); /* -e /dev/stdout  works */
#else
	  fprintf(stderr, "stderr redirection not possible without unistd.h\n");
#endif           
	}
	else if (!freopen(s1, "w", stderr)) {
	  fprintf(stderr, "stderr redirection to %s failed\n", s1);
	}
	break;
      case 'p': /* database path */
	job->cfg.db_path=s1; break;
      case 'o': /* output file */
	if (s1[0] == '-' && s1[1] == '\0') { } /* default = stdout */
	else if (!freopen(s1, "w", stdout)) {
	  fprintf(stderr, "stdout redirection to %s failed\n", s1);
	}
	break;
      case 'f': /* output format */
        if (strcmp(s1, "ISO8859_1") == 0) job->cfg.out_format=ISO8859_1; else
        if (strcmp(s1, "TeX")       == 0) job->cfg.out_format=TeX; else
        if (strcmp(s1, "HTML")      == 0) job->cfg.out_format=HTML; else
        if (strcmp(s1, "XML")       == 0) job->cfg.out_format=XML; else
        if (strcmp(s1, "SGML")      == 0) job->cfg.out_format=SGML; else
        if (strcmp(s1, "UTF8")      == 0) job->cfg.out_format=UTF8; else
        if (strcmp(s1, "ASCII")     == 0) job->cfg.out_format=ASCII; else
        fprintf(stderr,"Warning: unknown format (-f %s)\n",s1);
        i++;
        break;
      case 'c': /* list of chars (_ = not recognized chars) */
	job->cfg.lc = s1; break;
      case 'C': /* char filter, default: NULL (all chars) */
        /* ToDo: UTF8 input, wchar */
	job->cfg.cfilter = s1; break;
      case 'd': /* dust size */
	job->cfg.dust_size = atoi(s1); break;
      case 'l': /* grey level 0<160<=255, 0 for autodetect */
	job->cfg.cs = atoi(s1); break;
      case 's': /* spacewidth/dots (0 = autodetect) */
	job->cfg.spc = atoi(s1); break;
      case 'v': /* verbose mode */
	job->cfg.verbose |= atoi(s1); break;
      case 'm': /* operation modes */
	job->cfg.mode |= atoi(s1); break;
      case 'n': /* numbers only */
	job->cfg.only_numbers = atoi(s1); break;
      case 'x': /* initialize progress output s1=fname */
	ini_progress(s1); break;
      case 'a': /* set certainty */
	job->cfg.certainty = atoi(s1); break;
      case 'u': /* output marker for unrecognized chars */
        job->cfg.unrec_marker = s1; break;
      default:
	fprintf(stderr, "ERROR: unknown option %s, use -h for help\n",
	        argv[i]);
	exit(1); /* 2018-09 add exit on error */
      } /* switch(argv[i][1]) */
      i += iskip; /* 2017-03 add optional -v33 for -v 33 */
      continue;
    }
    else /* argument can be filename v0.2.5 */ if (argv[i][0] != '-'
						   || argv[i][1] == '\0' ) {
      job->src.fname = argv[i];
    }
  }
}

static void mark_start(job_t *job) {
  assert(job);

  if (job->cfg.verbose) {
    out_version(0);
    /* insert some helpful info for support */
    /* predefined macros:  gcc -dM -E - </dev/null # __unix __linux */
    /* ... but __DATE__ causes nonreproducable builds */
    fprintf(stderr, "# compiled: " /*__DATE__*/
#if   defined(__GNUC__)  /* can be supported by the author */
     " GNUC "
#elif defined(__TINYC__)  /* can be supported by the author */
     " TINYC "
#else
     " non-GNUC "  /* untested, pleasse report bugs + patches */
#endif
#if defined(__VERSION__)
    __VERSION__  /* gcc +clang */ 
#endif
#if   defined(__linux)
    " linux"     /* supported */
#else
    " non-linux" /* untested, please send bugs + patches */
#endif
       "\n");
    fprintf(stderr,
      "# options are: -l %d -s %d -v %d -c %s -m %d -d %d -n %d -a %d -C \"%s\"\n",
      job->cfg.cs, job->cfg.spc, job->cfg.verbose, job->cfg.lc, job->cfg.mode, 
      job->cfg.dust_size, job->cfg.only_numbers, job->cfg.certainty,
      job->cfg.cfilter);
    fprintf(stderr, "# file: %s\n", job->src.fname);
#ifdef USE_UNICODE
    fprintf(stderr,"# using unicode\n");
#endif
#ifdef HAVE_GETTIMEOFDAY
    gettimeofday(&job->tmp.init_time, NULL);
#endif
  }
}

static void mark_end(job_t *job) {
  assert(job);

#ifdef HAVE_GETTIMEOFDAY
  /* show elapsed time */
  if (job->cfg.verbose) {
    struct timeval end, result;
    gettimeofday(&end, NULL);
    timeval_subtract(&result, &end, &job->tmp.init_time);
    fprintf(stderr,"Elapsed time: %d:%02d:%3.3f.\n", (int)result.tv_sec/60,
	(int)result.tv_sec%60, (float)result.tv_usec/1000);
  }
#endif
}

static int read_picture(job_t *job) {
  int rc=0;
  assert(job);

  if (strstr(job->src.fname, ".pcx"))
    /* readpcx will exit on error (security) */
    readpcx(job->src.fname, &job->src.p, job->cfg.verbose);
  else
    rc=readpgm(job->src.fname, &job->src.p, job->cfg.verbose);
  return rc; /* 1 for multiple images, -1 on error, 0 else */
}

/* subject of change, we need more output for XML (ToDo) */
void print_output(job_t *job) {
  int linecounter = 0;
  const char *line;

  assert(job);
        
/* TODO: replace getTextLine-loop(line) by output_text() (libs have to use pipes)
   simplify code 2010-09-26
*/
  linecounter = 0;
  line = getTextLine(&(job->res.linelist), linecounter++);
  while (line) {
    /* notice: decode() is shiftet to getTextLine since 0.38 */
    fputs(line, stdout);
    if (job->cfg.out_format==HTML) fputs("<br />",stdout);
    if (job->cfg.out_format!=XML)  fputc('\n', stdout);
    line = getTextLine(&(job->res.linelist), linecounter++);
  }
  free_textlines(&(job->res.linelist));
}

/* FIXME jb: remove JOB; renamed to OCR_JOB 2010-09-26 */
job_t *OCR_JOB;


/* -------------------------------------------------------------
// ------   MAIN - replace this by your own aplication! 
// ------------------------------------------------------------- */
int main(int argn, char *argv[]) {
  int multipnm=1;
  job_t job1, *job; /* fixme, dont want global variables for lib */
  job=OCR_JOB=&job1;

  setvbuf(stdout, (char *) NULL, _IONBF, 0);	/* not buffered */

  job_init(job); /* init fname, db, cfg */

  process_arguments(job, argn, argv);
  
  /* load character data base (JS1002: now outside pgm2asc) */
  if ( job->cfg.mode & 2 ) /* check for db-option flag */
    load_db(job);
    /* load_db uses readpnm() and would conflict with multi images */
        
  while (multipnm==1) { /* multi-image loop */

    job_init_image(job); /* single image, ini res.box+line-list, tmp */

    mark_start(job);

    multipnm = read_picture(job);
    /* separation of main and rest for using as lib
       this will be changed later => introduction of set_option()
       for better communication to the engine  */
    if (multipnm<0) break; /* read error */
  
    /* call main loop */
    pgm2asc(job);

    mark_end(job);
  
    print_output(job);

    job_free_image(job);

  }

  return ((multipnm<0)?multipnm:0); /* -1=255 on error, 0 ok */
}
